das/rest_server_list.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package das
+
+import (
+	"bufio"
+	"context"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const initialMaxRecurseDepth uint16 = 8
+
+// RestfulServerURLsFromList reads a list of Restful server URLs from a remote URL.
+// The contents at the remote URL are parsed into a series of whitespace-separated words.
+// Each word is interpreted as the URL of a Restful server, except that if a word is "LIST"
+//    (case-insensitive) then the following word is interpreted as the URL of another list,
+//    which is recursively fetched. The depth of recursion is limited to initialMaxRecurseDepth.
+func RestfulServerURLsFromList(ctx context.Context, listUrl string) ([]string, error) {
+	client := &http.Client{}
+	urls, err := restfulServerURLsFromList(ctx, client, listUrl, initialMaxRecurseDepth)
+	if err != nil {
+		return nil, err
+	}
+
+	// deduplicate the list of URL strings
+	seen := make(map[string]bool)
+	dedupedUrls := []string{}
+	for _, url := range urls {
+		if !seen[url] {
+			seen[url] = true
+			dedupedUrls = append(dedupedUrls, url)
+		}
+	}
+
+	return dedupedUrls, nil
+}
+
+func restfulServerURLsFromList(ctx context.Context, client *http.Client, listUrl string, maxRecurseDepth uint16) ([]string, error) {
+	urls := []string{}
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, listUrl, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+	scanner := bufio.NewScanner(resp.Body)
+	scanner.Split(bufio.ScanWords)
+	for scanner.Scan() {
+		word := scanner.Text()
+		if strings.ToLower(word) == "list" {
+			if maxRecurseDepth > 0 && scanner.Scan() {
+				word = scanner.Text()
+				subUrls, err := restfulServerURLsFromList(ctx, client, word, maxRecurseDepth-1)
```

Comment:
Do you think we should worry about cycle detection?

Reply:
Added cycle detection.